### DISCUSSION TITLE
useCellRangeSelection - Working on this hook, Ability to select cell ranges and single select (UI demo)

### DISCUSSION THREAD

**Discussion by 07harish** [TanStack/table#2476] [Ideas]
# Multi-cell range selection and single-cell select feature

Hi @tannerlinsley,  Implemented this new hook, Will this be a good addition to the library?

 @gargroh helped me to come up with this idea.

![useCellRangeSelection](https://user-images.githubusercontent.com/27046938/85893927-67951600-b811-11ea-8a3e-3ca4de85f724.gif)
 
### Usage:

- Multiple cell range selections can be done using `Ctrl + range selection (click and drag)`. First range selection can be done without `Ctrl`
- Cell selections (and unselection) can be done using `Ctrl + click`

### Compatibility:
This feature works well with `useColumnOrder`. With `useGroupBy`, it selects all cells within the range (including empty cells when expanded)

### Few details on the implementation:
-  added cell Id's
- `state.selectedCells` Is a map of `cell.id = true`, Same as `selectedRowIds`
- I have added `instance.cellsById`,  actions for `start, selecting, end` on an `instance`

### Few things I have not considered adding (will work on it, if this is required): 
 - Unselecting cell ranges
 - Ability to represent like layer for overlapping range selections, but I am thinking if this can be done by keeping count
  `state.selectedCells : {cell.id :  count, ...}
`

I am new to contributing to open source, Need your thoughts and feedback.
 If this looks good, I have some refactoring left and I will create a PR.. Thanks :)

**Comment by etodanik** [comment]
We could definitely use something like this at our company. I'd love to help out building this in whatever way necessary.

**Comment by 07harish** [comment]
Update: @israelidanny @tannerlinsley - This PR is merged in [react-table-plugins ](https://github.com/gargroh/react-table-plugins/) 🎉 

 Check [Documentation](https://github.com/gargroh/react-table-plugins/blob/master/docs/useCellRangeSelection.md) and [codesandbox](https://codesandbox.io/s/github/gargroh/react-table-plugins/tree/master/examples/cell-range-selection)

Let me know, what do you think and if it's a good addition here, will raise a PR

**Reply by etodanik** [reply]
You're a legend! Will check it.

**Comment by tannerlinsley** [comment]
I would like to see this integrated with v8 (the `next` branch). Wanna take a whack at it there? You'll need to look at how other plugins have changed to fit the new architecture.

**Comment by 07harish** [comment]
Sure, I will have a look at the new architecture and will try to implement it there.

**Comment by 07harish** [comment]
@tannerlinsley If you have time, Can you please look at the implementation and can review the code? So that we are following the right direction:

 [Code link](https://github.com/gargroh/react-table-plugins/blob/master/src/plugins/useCellRangeSelection.js),
[Codesandbox](https://codesandbox.io/s/github/gargroh/react-table-plugins/tree/master/examples/cell-range-selection) and [Docs](https://github.com/gargroh/react-table-plugins/blob/master/docs/useCellRangeSelection.md)

**Comment by tannerlinsley** [comment]
This looks awesome. And yes, the architecture looks pretty good. Things might change a *little* bit with v8, so I'll keep you posted.

**Comment by gopal-abbineni** [comment]
@tannerlinsley @07harish I think default click cell selection will be more convenient to select single cell instead of doing ctrl+click and also is there any plans for shift click selection between two cells?

**Comment by bolam-o** [comment]
@07harish this plugin is not support typescript?

**Comment by andresgutgon** [comment]
Hi, I'm trying to port this to React Table v8. 
The part I'm struggling is where to do this hash [cellsById](https://github.com/gargroh/react-table-plugins/blob/master/src/plugins/useCellRangeSelection.js#L225-L232)

What would be the best place?

**Reply by joshkay** [reply]
Did you make any progress on this?  Looking at setting it up myself.

**Reply by ambid17** [reply]
bump, I am curious if you ever got this working, I am also struggling to get this to work in v8 :D

**Comment by joshkay** [comment]
> bump, I am curious if you ever got this working, I am also struggling to get this to work in v8 :D

Ended up writing my own hook in the end.  
[Here's a gist](https://gist.github.com/joshkay/fc8bab0561583dd48fecce93022fc7a2) with the hook & the table component I'm using.
